In [109]:
## Importing all necessary tools:
import pandas as pd
import numpy as np
pd.options.display.max_columns = None

import warnings
warnings.filterwarnings("ignore")        ## Hides all warnings

In [111]:
df = pd.read_csv("Housing_Listings_all_records_(numbers)_FINAL_Cleaned_FE.csv", index_col = "Unnamed: 0")
df.head(2)

Flat     Sector                            Address  \
0  2 BHK Flat  Sector 86  Pyramid Elite, Sector 86, Gurgaon   
1  2 BHK Flat  Sector 86  Pyramid Elite, Sector 86, Gurgaon   

                      Seller_Builder                     EMI  EMI_in_rupees  \
0  Pyramid Infratech Private Limited  EMI starts at ₹36.01 K          36010   
1  Pyramid Infratech Private Limited  EMI starts at ₹33.36 K          33360   

  Built_Up_Area  Built_up_area_in_sqft       Avg_Price  \
0     593 sq.ft                    593  ₹11.47 K/sq.ft   
1     690 sq.ft                    690   ₹9.13 K/sq.ft   

   Avg_price_rupee_per_sqft Age_of_property Age_of_property_in_years  \
0                   11470.0     1 Years Old                        1   
1                    9130.0     1 Years Old                        1   

  Possession_status     Floor Floor_number Building_height       Facing  \
0     Ready to move  12 of 14           12              14  East facing   
1     Ready to move   4 of 15            4              15            -   

       Furnishing        Society Brokerage  Brokerage_in_rupees   Price  \
0  Semi Furnished  Pyramid Elite    68,000                68000  68.0 L   
1     Unfurnished  Pyramid Elite    63,000                63000  63.0 L   

   Price_in_rupees  Bedrooms Bathrooms         Parking Covered_parking  \
0          6800000         2         2  1 Open Parking               0   
1          6300000         2         2      No Parking               0   

  Open_parking Balcony             Advertised  \
0            1       1  More than a month ago   
1            0       1            17 days ago   

                                           Amenities  \
0  ['Lift', 'Power Backup', 'Garden', 'Sports', '...   
1                                                NaN   

                                    Nearby_landmarks  \
0  [['School', "St. Xavier's High School"], ['Hos...   
1  [['School', "St. Xavier's High School"], ['Hos...   

                                    Prop_description  \
0  Looking for a 2 BHK Flat for sale in Gurgaon? ...   
1  Best 2 BHK Flat for modern-day lifestyle is no...   

                                                Link  Sector_Amenity_Score  
0  https://housing.com/in/buy/resale/page/1761033...                   2.0  
1  https://housing.com/in/buy/resale/page/1681859...                   2.0

## **Useful Features are:**
1. **Flat** - Tells Flat type.
2. **Sector_Locality** - Tells Sector or smaller locality of Gurugram where the property falls.
3. **Locality** - Tells broader Loaclity of Gurugram where the property falls.
4. **Built_up_area_in_sqft** - Tells the total Built-up Area of Property.
5. **Sector_Locality_Price_Density** - Tells the prevalent Price Density of the region. [Based on Historical data!]}
6. **Age_Category** - Tells the Age Category of Property (in Years).
7. **Floor_Category** - Tells the relative position of FLoor in the Building.
8. **Building_Category** - Tells the height category of Building.
11. **Furnishing** - Tells the Furnishing status of Property.
12. **Price_in_rupees** - Tells property Price in rupees.
14. **Bedrooms** - Tells the number of Bedrooms in the Property.
15. **Bathrooms** - Tells the total number of Bathrooms in the Property.
16. **Balcony** - Tells the number of Balconies in the Property.
17. **Open_Parking** - Tells the Availability of Open Parking.
18. **Covered_Parking** - Tells the Availability of Closed Parking.
19. **Sector_Amenity_Score** - Tells Amenity Score of a Sector out of 10. 

**NOTE:** 
- All 1.5, 2.5, 3.5 and 4.5 BHK Flats have 2, 3, 4 and 5 Bedrooms respectively.
- We would not be using **Facing** feature reasons- Too many missing entries (1919) and very small Price Variance is explained.
- Also not using the following features to evade **Data Leakage**: ***Price Density, EMI and Brokerage Cost***.

<br><br>

---
## `I.` Creating New Column - "Sector_Locality" - adding Locality name for NULL Sector values.

In [116]:
## Creating new "Sector_Locality" feature from Address:
def func_1(x):
    if x.find('Sector ') != -1:
        x = x.split(', ')
        for i in range(len(x)):
            if x[i][0:7] == 'Sector ':
                temp = x[i]
                res = temp.split(" ", 2)[0] + " " + temp.split(" ", 2)[1]
                return res
    else:
        temp = x.split(", ")[-2]
        return temp
        
value = df['Address'].apply(func_1)
value[3302] = 'Sector 15'    ## Anomaly

df.insert(1, 'Sector_Locality', value, allow_duplicates=False)
df.head(3)

Flat Sector_Locality      Sector  \
0  2 BHK Flat       Sector 86   Sector 86   
1  2 BHK Flat       Sector 86   Sector 86   
2  2 BHK Flat      Sector 108  Sector 108   

                                       Address  \
0            Pyramid Elite, Sector 86, Gurgaon   
1            Pyramid Elite, Sector 86, Gurgaon   
2  Experion The Heartsong, Sector 108, Gurgaon   

                      Seller_Builder                     EMI  EMI_in_rupees  \
0  Pyramid Infratech Private Limited  EMI starts at ₹36.01 K          36010   
1  Pyramid Infratech Private Limited  EMI starts at ₹33.36 K          33360   
2                Experion Developers  EMI starts at ₹74.47 K          74470   

  Built_Up_Area  Built_up_area_in_sqft       Avg_Price  \
0     593 sq.ft                    593  ₹11.47 K/sq.ft   
1     690 sq.ft                    690   ₹9.13 K/sq.ft   
2    1000 sq.ft                   1000     ₹15 K/sq.ft   

   Avg_price_rupee_per_sqft Age_of_property Age_of_property_in_years  \
0                   11470.0     1 Years Old                        1   
1                    9130.0     1 Years Old                        1   
2                   15000.0      2 Year Old                        2   

  Possession_status     Floor Floor_number Building_height             Facing  \
0     Ready to move  12 of 14           12              14        East facing   
1     Ready to move   4 of 15            4              15                  -   
2     Ready to move   9 of 26            9              26  North-East facing   

       Furnishing                 Society Brokerage  Brokerage_in_rupees  \
0  Semi Furnished           Pyramid Elite    68,000                68000   
1     Unfurnished           Pyramid Elite    63,000                63000   
2  Semi Furnished  Experion The Heartsong  1.5 Lacs               150000   

    Price  Price_in_rupees  Bedrooms Bathrooms                       Parking  \
0  68.0 L          6800000         2         2                1 Open Parking   
1  63.0 L          6300000         2         2                    No Parking   
2  1.5 Cr         15000000         2         2  1 Covered and 1 Open Parking   

  Covered_parking Open_parking Balcony             Advertised  \
0               0            1       1  More than a month ago   
1               0            0       1            17 days ago   
2               1            1       3  More than a month ago   

                                           Amenities  \
0  ['Lift', 'Power Backup', 'Garden', 'Sports', '...   
1                                                NaN   
2  ['Amphitheater', 'Cricket Pitch', 'Gazebo', 'S...   

                                    Nearby_landmarks  \
0  [['School', "St. Xavier's High School"], ['Hos...   
1  [['School', "St. Xavier's High School"], ['Hos...   
2  [['School', 'The Shikshiyan School'], ['Hospit...   

                                    Prop_description  \
0  Looking for a 2 BHK Flat for sale in Gurgaon? ...   
1  Best 2 BHK Flat for modern-day lifestyle is no...   
2  2 BHK Flat for sale in Sector 108, Gurgaon - c...   

                                                Link  Sector_Amenity_Score  
0  https://housing.com/in/buy/resale/page/1761033...                   2.0  
1  https://housing.com/in/buy/resale/page/1681859...                   2.0  
2  https://housing.com/in/buy/resale/page/1767484...                   0.0

<br><br>

---
## `II.` Creating New Column - "Locality" using Address - dividing entire Gurugram into Wider Localities.

In [119]:
Other_localities = {'Ashok Vihar Phase 2': 'New Gurugram',
 'Civil Lines': 'Old Gurgaon',
 'DLF Phase 1': 'MG Road & Central Gurugram',
 'DLF Phase 2': 'MG Road & Central Gurugram',
 'DLF Phase 3': 'MG Road & Central Gurugram',
 'DLF Phase 4': 'MG Road & Central Gurugram',
 'DLF Phase 5': 'Golf Course Road & Extension',
 'Dhunela': 'Dhunela',
 'Gwal Pahari': 'Gwal Pahari',
 'Kadarpur': 'Kadarpur',
 'Manesar': 'Old Gurgaon',
 'Nurpur Jharsa': 'Southern Peripheral Road (SPR) Belt',
 'Palam Farms': 'Old Gurgaon',
 'Palam Vihar': 'Old Gurgaon',
 'Palam Vihar Extension': 'Old Gurgaon',
 'Patel Nagar': 'Old Gurgaon',
 'Sadhrana Bamripur': 'Sadhrana Bamripur',
 'Southern Peripheral Road': 'Southern Peripheral Road (SPR) Belt',
 'Sushant Lok Phase 1': 'MG Road & Central Gurugram',
 'Tulip Violet Society': 'Southern Peripheral Road (SPR) Belt'}

## Using Sector-wise Geo Data to get Locality of Sector:
sec_location = pd.read_csv('sector_geo_centroid_data.csv', index_col = 'Unnamed: 0')
temp_dict = sec_location[['Sector', 'Locality']].set_index('Sector').to_dict()['Locality']

temp_dict.update(Other_localities)

## ------------------------------------------------------------------------------------------------------
## Creating and Adding "Locality" feature to dataframe:
loc_series = df['Sector_Locality'].replace(temp_dict)

df.insert(2, 'Locality', loc_series, allow_duplicates=False)
df.head(3)

Flat Sector_Locality                Locality      Sector  \
0  2 BHK Flat       Sector 86            New Gurugram   Sector 86   
1  2 BHK Flat       Sector 86            New Gurugram   Sector 86   
2  2 BHK Flat      Sector 108  Dwarka Expressway Belt  Sector 108   

                                       Address  \
0            Pyramid Elite, Sector 86, Gurgaon   
1            Pyramid Elite, Sector 86, Gurgaon   
2  Experion The Heartsong, Sector 108, Gurgaon   

                      Seller_Builder                     EMI  EMI_in_rupees  \
0  Pyramid Infratech Private Limited  EMI starts at ₹36.01 K          36010   
1  Pyramid Infratech Private Limited  EMI starts at ₹33.36 K          33360   
2                Experion Developers  EMI starts at ₹74.47 K          74470   

  Built_Up_Area  Built_up_area_in_sqft       Avg_Price  \
0     593 sq.ft                    593  ₹11.47 K/sq.ft   
1     690 sq.ft                    690   ₹9.13 K/sq.ft   
2    1000 sq.ft                   1000     ₹15 K/sq.ft   

   Avg_price_rupee_per_sqft Age_of_property Age_of_property_in_years  \
0                   11470.0     1 Years Old                        1   
1                    9130.0     1 Years Old                        1   
2                   15000.0      2 Year Old                        2   

  Possession_status     Floor Floor_number Building_height             Facing  \
0     Ready to move  12 of 14           12              14        East facing   
1     Ready to move   4 of 15            4              15                  -   
2     Ready to move   9 of 26            9              26  North-East facing   

       Furnishing                 Society Brokerage  Brokerage_in_rupees  \
0  Semi Furnished           Pyramid Elite    68,000                68000   
1     Unfurnished           Pyramid Elite    63,000                63000   
2  Semi Furnished  Experion The Heartsong  1.5 Lacs               150000   

    Price  Price_in_rupees  Bedrooms Bathrooms                       Parking  \
0  68.0 L          6800000         2         2                1 Open Parking   
1  63.0 L          6300000         2         2                    No Parking   
2  1.5 Cr         15000000         2         2  1 Covered and 1 Open Parking   

  Covered_parking Open_parking Balcony             Advertised  \
0               0            1       1  More than a month ago   
1               0            0       1            17 days ago   
2               1            1       3  More than a month ago   

                                           Amenities  \
0  ['Lift', 'Power Backup', 'Garden', 'Sports', '...   
1                                                NaN   
2  ['Amphitheater', 'Cricket Pitch', 'Gazebo', 'S...   

                                    Nearby_landmarks  \
0  [['School', "St. Xavier's High School"], ['Hos...   
1  [['School', "St. Xavier's High School"], ['Hos...   
2  [['School', 'The Shikshiyan School'], ['Hospit...   

                                    Prop_description  \
0  Looking for a 2 BHK Flat for sale in Gurgaon? ...   
1  Best 2 BHK Flat for modern-day lifestyle is no...   
2  2 BHK Flat for sale in Sector 108, Gurgaon - c...   

                                                Link  Sector_Amenity_Score  
0  https://housing.com/in/buy/resale/page/1761033...                   2.0  
1  https://housing.com/in/buy/resale/page/1681859...                   2.0  
2  https://housing.com/in/buy/resale/page/1767484...                   0.0

<br><br>

---
## `III.` Creating New Column - "Age_Category" using Age_of_property_in_years - dividing properties into 4 Age Categories.

In [121]:
## Classifying Properties based on Age of the Property in Years:
def func_new(data):
    if data == '-' or int(data) <= 5:    ## Temporarily putting all the NAN entries as "Newer Properties"...
        return "Newer Property"
    elif int(data) > 5 and int(data) <= 10:
        return "Mid Age Property"
    elif int(data) > 10 and int(data) <= 15:
        return "Slightly Older"
    elif int(data) > 15:
        return "Older Property"


Age_cate = df['Age_of_property_in_years'].apply(func_new)

df.insert(14, 'Age_Category', Age_cate, allow_duplicates=False)
df.head(3)

Flat Sector_Locality                Locality      Sector  \
0  2 BHK Flat       Sector 86            New Gurugram   Sector 86   
1  2 BHK Flat       Sector 86            New Gurugram   Sector 86   
2  2 BHK Flat      Sector 108  Dwarka Expressway Belt  Sector 108   

                                       Address  \
0            Pyramid Elite, Sector 86, Gurgaon   
1            Pyramid Elite, Sector 86, Gurgaon   
2  Experion The Heartsong, Sector 108, Gurgaon   

                      Seller_Builder                     EMI  EMI_in_rupees  \
0  Pyramid Infratech Private Limited  EMI starts at ₹36.01 K          36010   
1  Pyramid Infratech Private Limited  EMI starts at ₹33.36 K          33360   
2                Experion Developers  EMI starts at ₹74.47 K          74470   

  Built_Up_Area  Built_up_area_in_sqft       Avg_Price  \
0     593 sq.ft                    593  ₹11.47 K/sq.ft   
1     690 sq.ft                    690   ₹9.13 K/sq.ft   
2    1000 sq.ft                   1000     ₹15 K/sq.ft   

   Avg_price_rupee_per_sqft Age_of_property Age_of_property_in_years  \
0                   11470.0     1 Years Old                        1   
1                    9130.0     1 Years Old                        1   
2                   15000.0      2 Year Old                        2   

     Age_Category Possession_status     Floor Floor_number Building_height  \
0  Newer Property     Ready to move  12 of 14           12              14   
1  Newer Property     Ready to move   4 of 15            4              15   
2  Newer Property     Ready to move   9 of 26            9              26   

              Facing      Furnishing                 Society Brokerage  \
0        East facing  Semi Furnished           Pyramid Elite    68,000   
1                  -     Unfurnished           Pyramid Elite    63,000   
2  North-East facing  Semi Furnished  Experion The Heartsong  1.5 Lacs   

   Brokerage_in_rupees   Price  Price_in_rupees  Bedrooms Bathrooms  \
0                68000  68.0 L          6800000         2         2   
1                63000  63.0 L          6300000         2         2   
2               150000  1.5 Cr         15000000         2         2   

                        Parking Covered_parking Open_parking Balcony  \
0                1 Open Parking               0            1       1   
1                    No Parking               0            0       1   
2  1 Covered and 1 Open Parking               1            1       3   

              Advertised                                          Amenities  \
0  More than a month ago  ['Lift', 'Power Backup', 'Garden', 'Sports', '...   
1            17 days ago                                                NaN   
2  More than a month ago  ['Amphitheater', 'Cricket Pitch', 'Gazebo', 'S...   

                                    Nearby_landmarks  \
0  [['School', "St. Xavier's High School"], ['Hos...   
1  [['School', "St. Xavier's High School"], ['Hos...   
2  [['School', 'The Shikshiyan School'], ['Hospit...   

                                    Prop_description  \
0  Looking for a 2 BHK Flat for sale in Gurgaon? ...   
1  Best 2 BHK Flat for modern-day lifestyle is no...   
2  2 BHK Flat for sale in Sector 108, Gurgaon - c...   

                                                Link  Sector_Amenity_Score  
0  https://housing.com/in/buy/resale/page/1761033...                   2.0  
1  https://housing.com/in/buy/resale/page/1681859...                   2.0  
2  https://housing.com/in/buy/resale/page/1767484...                   0.0

### `Revising the Age_Category feature by imputing the NAN entries with better values:`

In [123]:
count = 0
def func_new_1(data):
    global count
    if data['Age_of_property_in_years'] == '-':
        if len(df[(df['Sector_Locality'] == data['Sector_Locality']) & (df['Age_of_property_in_years'] != '-')]['Age_Category'].mode().values) < 1:
            count += 1
            return df[(df['Sector_Locality'] == data['Sector_Locality'])]['Age_Category'].mode().values[0]
        else:
            return df[(df['Sector_Locality'] == data['Sector_Locality']) & (df['Age_of_property_in_years'] != '-')]['Age_Category'].mode().values[0]
    else:
        return data['Age_Category']
        
   
Age_cate_revised = df.apply(func_new_1, axis=1)
print("Number of Properties whose Sector or Locality does not have any Property with Valid Property Age : ", count)
df.drop(columns = ['Age_Category'], inplace = True)   ## Dropping old "Age_Category" Column....

df.insert(14, 'Age_Category', Age_cate, allow_duplicates=False)
df.head(2)

Number of Properties whose Sector or Locality does not have any Property with Valid Property Age :  14


Flat Sector_Locality      Locality     Sector  \
0  2 BHK Flat       Sector 86  New Gurugram  Sector 86   
1  2 BHK Flat       Sector 86  New Gurugram  Sector 86   

                             Address                     Seller_Builder  \
0  Pyramid Elite, Sector 86, Gurgaon  Pyramid Infratech Private Limited   
1  Pyramid Elite, Sector 86, Gurgaon  Pyramid Infratech Private Limited   

                      EMI  EMI_in_rupees Built_Up_Area  Built_up_area_in_sqft  \
0  EMI starts at ₹36.01 K          36010     593 sq.ft                    593   
1  EMI starts at ₹33.36 K          33360     690 sq.ft                    690   

        Avg_Price  Avg_price_rupee_per_sqft Age_of_property  \
0  ₹11.47 K/sq.ft                   11470.0     1 Years Old   
1   ₹9.13 K/sq.ft                    9130.0     1 Years Old   

  Age_of_property_in_years    Age_Category Possession_status     Floor  \
0                        1  Newer Property     Ready to move  12 of 14   
1                        1  Newer Property     Ready to move   4 of 15   

  Floor_number Building_height       Facing      Furnishing        Society  \
0           12              14  East facing  Semi Furnished  Pyramid Elite   
1            4              15            -     Unfurnished  Pyramid Elite   

  Brokerage  Brokerage_in_rupees   Price  Price_in_rupees  Bedrooms Bathrooms  \
0    68,000                68000  68.0 L          6800000         2         2   
1    63,000                63000  63.0 L          6300000         2         2   

          Parking Covered_parking Open_parking Balcony             Advertised  \
0  1 Open Parking               0            1       1  More than a month ago   
1      No Parking               0            0       1            17 days ago   

                                           Amenities  \
0  ['Lift', 'Power Backup', 'Garden', 'Sports', '...   
1                                                NaN   

                                    Nearby_landmarks  \
0  [['School', "St. Xavier's High School"], ['Hos...   
1  [['School', "St. Xavier's High School"], ['Hos...   

                                    Prop_description  \
0  Looking for a 2 BHK Flat for sale in Gurgaon? ...   
1  Best 2 BHK Flat for modern-day lifestyle is no...   

                                                Link  Sector_Amenity_Score  
0  https://housing.com/in/buy/resale/page/1761033...                   2.0  
1  https://housing.com/in/buy/resale/page/1681859...                   2.0

<br><br>

---
## `IV.` Creating New Column - "Floor_Category" & "Building_Height_Category" using 'Floor_number' & 'Building_height' - classifying properties based on relative position of property in the Building and Building Height.

In [125]:
print("Number of Properties with Floor information missing are : ", df[df['Floor'] == '-'].shape[0])
print("Number of properties with Floor & Balcony information missing : ", 
                                                            df[(df['Floor'] == '-') & (df['Balcony'] == '-')] .shape[0])

Number of Properties with Floor information missing are :  347
Number of properties with Floor & Balcony information missing :  186


### **`WAY FORWARD :`**
- We can simply put the Floor Number as 0 or Ground Floor, and
- Building Height as 1
- Also since, there are 186 Properties with Balcony Number and Floor information missing.
- They are highly likely to be Ground Floor Properties where Balcony count does not matter.

In [127]:
## Imputing missing Floor Number as 'Ground Floor' & Building Height as '1':
df['Floor_number'].replace({'-': 'Ground Floor'}, inplace = True)
df['Building_height'].replace({'-': '1'}, inplace = True)

In [128]:
## Classifying Properties based on relative position of Floor in the Building & Building Height:
## Lets Categorise Buildings based on their Total Floor Heights:
thresh_1 = np.percentile(df['Building_height'].astype('int32'), 33.33).astype('int32')   ## i.e., 14
thresh_2 = np.percentile(df['Building_height'].astype('int32'), 66.66).astype('int32')   ## i.e., 20


def func_new(data, thresh_1, thresh_2):
    height = data['Building_height']
    floor = data['Floor_number']
    floor_cate = None
    building_cate = None
    
    ## Floor Category:
    if floor <= (1/3) * height:
        floor_cate = 'Lower Floor'                                 ## Floor in lower 1/3rd part
    elif floor > (1/3) * height and floor <= (2/3) * height:
        floor_cate = 'Middle Floor'                                ## Floor in middle 1/3rd part
    else:
        floor_cate = 'Upper Floor'                                 ## Floor in upper 1/3rd part
    
    ## Building Category:
    if height <= thresh_1:
        building_cate = "Shorter Building"                         ## Height between 1 to 14
    elif height > thresh_1 and height <= thresh_2:
        building_cate = "Medium Building"                          ## Height between 14 to 20
    else:
        building_cate = "Taller Building"                          ## Height greater than 20

    return pd.Series([floor_cate, building_cate], index = ['Floor_number', 'Building_height'])

data = df[['Floor_number', 'Building_height']]
data['Floor_number'] = data['Floor_number'].replace({'Ground Floor': '0'}).astype('int32')
data['Building_height'] = data['Building_height'].astype('int32')

Floor_number_Building_height = data.apply(func_new, thresh_1=thresh_1, thresh_2=thresh_2, axis = 1)
Floor_number_Building_height

df.insert(19, 'Floor_Category', Floor_number_Building_height['Floor_number'], allow_duplicates=False)
df.insert(20, 'Building_Height_Category', Floor_number_Building_height['Building_height'], allow_duplicates=False)

df.head(3)

Flat Sector_Locality                Locality      Sector  \
0  2 BHK Flat       Sector 86            New Gurugram   Sector 86   
1  2 BHK Flat       Sector 86            New Gurugram   Sector 86   
2  2 BHK Flat      Sector 108  Dwarka Expressway Belt  Sector 108   

                                       Address  \
0            Pyramid Elite, Sector 86, Gurgaon   
1            Pyramid Elite, Sector 86, Gurgaon   
2  Experion The Heartsong, Sector 108, Gurgaon   

                      Seller_Builder                     EMI  EMI_in_rupees  \
0  Pyramid Infratech Private Limited  EMI starts at ₹36.01 K          36010   
1  Pyramid Infratech Private Limited  EMI starts at ₹33.36 K          33360   
2                Experion Developers  EMI starts at ₹74.47 K          74470   

  Built_Up_Area  Built_up_area_in_sqft       Avg_Price  \
0     593 sq.ft                    593  ₹11.47 K/sq.ft   
1     690 sq.ft                    690   ₹9.13 K/sq.ft   
2    1000 sq.ft                   1000     ₹15 K/sq.ft   

   Avg_price_rupee_per_sqft Age_of_property Age_of_property_in_years  \
0                   11470.0     1 Years Old                        1   
1                    9130.0     1 Years Old                        1   
2                   15000.0      2 Year Old                        2   

     Age_Category Possession_status     Floor Floor_number Building_height  \
0  Newer Property     Ready to move  12 of 14           12              14   
1  Newer Property     Ready to move   4 of 15            4              15   
2  Newer Property     Ready to move   9 of 26            9              26   

  Floor_Category Building_Height_Category             Facing      Furnishing  \
0    Upper Floor         Shorter Building        East facing  Semi Furnished   
1    Lower Floor          Medium Building                  -     Unfurnished   
2   Middle Floor          Taller Building  North-East facing  Semi Furnished   

                  Society Brokerage  Brokerage_in_rupees   Price  \
0           Pyramid Elite    68,000                68000  68.0 L   
1           Pyramid Elite    63,000                63000  63.0 L   
2  Experion The Heartsong  1.5 Lacs               150000  1.5 Cr   

   Price_in_rupees  Bedrooms Bathrooms                       Parking  \
0          6800000         2         2                1 Open Parking   
1          6300000         2         2                    No Parking   
2         15000000         2         2  1 Covered and 1 Open Parking   

  Covered_parking Open_parking Balcony             Advertised  \
0               0            1       1  More than a month ago   
1               0            0       1            17 days ago   
2               1            1       3  More than a month ago   

                                           Amenities  \
0  ['Lift', 'Power Backup', 'Garden', 'Sports', '...   
1                                                NaN   
2  ['Amphitheater', 'Cricket Pitch', 'Gazebo', 'S...   

                                    Nearby_landmarks  \
0  [['School', "St. Xavier's High School"], ['Hos...   
1  [['School', "St. Xavier's High School"], ['Hos...   
2  [['School', 'The Shikshiyan School'], ['Hospit...   

                                    Prop_description  \
0  Looking for a 2 BHK Flat for sale in Gurgaon? ...   
1  Best 2 BHK Flat for modern-day lifestyle is no...   
2  2 BHK Flat for sale in Sector 108, Gurgaon - c...   

                                                Link  Sector_Amenity_Score  
0  https://housing.com/in/buy/resale/page/1761033...                   2.0  
1  https://housing.com/in/buy/resale/page/1681859...                   2.0  
2  https://housing.com/in/buy/resale/page/1767484...                   0.0

<br><br>

---
## `V.` Creating New Column - "Sector_Locality_Price_Density" using Price Density Data for each Sector-Locality - this will record the prevalent Price Density of the Sector-Locality.

- 📊 **Median Price Density of Sector or Locality in Price Prediction Models:**
- **USEFULNESS:** Median Price Density (e.g., median price per sq.ft of an area/sector) is a **very strong location-level signal**.
    - It captures the **prevailing market sentiment, affordability, and positioning** of that area.
    - Including it can significantly boost predictive power, especially in heterogeneous markets like Gurugram, where location is the primary driver of value.<br><br>

* **Risk of Data Leakage:** It depends on **how the feature is constructed**:<br>
  - ✅ **Safe:** If Median Price Density is computed from **historical or external market data** (independent of your training dataset), it’s fine to use—it acts like a sector-level benchmark.
  - ⚠️ **Leakage Risk:** If you calculate Median Price Density directly from your **training dataset’s target variable (property prices)** and then feed it back as a feature, you are inadvertently leaking information about the very variable you’re trying to predict. **This will inflate model accuracy unrealistically**.<br><br>

* **Best Practice:** <br>
  - Compute Median Price Density from **external market research sources** (e.g., CREDAI, JLL, Knight Frank reports, or government circle rate data).
  - Or, if using your dataset, calculate median prices **only on the training set** and **apply them as sector-level features** to **test/validation sets** (no peeking ahead).


✅ **IN SHORT:** Median Price Density is a **powerful feature** for price prediction **but must be engineered carefully** to **avoid leakage**. The key is ensuring it represents **independent location intelligence**, not a repackaged version of your target variable.


### **`Let's Apply it safely:`**
- We'll split the dataset into 5500 records for training (~ 79%) and 1500 for testing (~ 21%).
- Then create this Feature using training dataset alone and use it as historical records while testing.
- This way we'll able to avoid Data Leakage.

In [131]:
## Code to be applied over Training data set:
## __________________________________________

## Finding Median Price Density of each  Sector or Locality:
Sec_loc_PD = df.groupby(by = 'Sector_Locality').agg({'Avg_price_rupee_per_sqft': 'median'
                                                    }).to_dict()['Avg_price_rupee_per_sqft']

values = df['Sector_Locality'].replace(Sec_loc_PD)
df.insert(6, "Sector_Locality_Price_Density", values, allow_duplicates = False)

<br><br>

---
## `VI.` Missing Value Imputation in "Sector_Amenity_Score" feature- the imputation will be based on minimum aggregated Score of the Broader Locality in which that Sector or Locality falls.

In [133]:
def func(data):
    if data['Sector'] != '-':
        return data['Sector_Amenity_Score']
    else:
        if data['Locality'] in ['New Gurugram', 'Dwarka Expressway Belt', 'Old Gurgaon',  'MG Road & Central Gurugram',
                                'Southern Peripheral Road (SPR) Belt', 'Golf Course Road & Extension']:
            sec_am_score = df[(df['Sector'] != '-') & 
                                            (df['Locality'] == data['Locality'])]['Sector_Amenity_Score'].min()
            return sec_am_score
        else:
            return 0

values = df.apply(func, axis = 1)
df['Sector_Amenity_Score'] = values

<br><br>

---
## `VII.` Missing Value Imputation in "Balcony" and "Bathrooms" features- the imputation will be based on data patterns and logic.

In [136]:
## For "Balcony" Feature:
def func(data):
    if data['Balcony'] != '-':     ## Do not change if Balcony value is not missing...
        return data['Balcony']
    elif data['Floor_number'] == 'Ground Floor':    ## Putting "No Balcony", if its Ground Floor property...
        return "No Balcony"
    else:
        possible_balcony = df[df['Flat'] == data['Flat']]['Balcony'].mode().values[0]   
        ## Putting most likely Balcony count as per the Flat Type... 
        return possible_balcony

values = df.apply(func, axis = 1)
df['Balcony'] = values.replace({'No Balcony': '0'}).astype('int32')

In [137]:
## For "Bathrooms" Feature:
def func(data):
    if data['Bathrooms'] != '-':     ## ## Do not change if Bathroom count is not missing...
        return data['Bathrooms']
    else:
        possible_bathrooms = df[df['Flat'] == data['Flat']]['Bathrooms'].mode().values[0] 
        return possible_bathrooms


values = df.apply(func, axis = 1)
df['Bathrooms'] = values.astype('int32')

<br><br>

---
## `VIII.` Missing Value Imputation in "Furnishing", "Open and Covered Parking" features- simply imputing the most common entry of the Feature

In [139]:
## For "Furnishing" Feature:
df['Furnishing'].replace({'-': df['Furnishing'].mode().values[0]}, inplace = True)

## For "Open_parking" & "Covered_parking" Feature:
df['Open_parking'].replace({'-': '0'}, inplace = True)
df['Covered_parking'].replace({'-': '0'}, inplace = True)

df['Open_parking'] = df['Open_parking'].astype('int32')
df['Covered_parking'] = df['Covered_parking'].astype('int32')

<br><br><br>

---
## **`Separating all Usefull Features of main DF:`** <i>{for Model Building}<i>

In [141]:
## Keeping only the usefull features:
data = df[['Flat', 'Sector_Locality', 'Locality', 'Built_up_area_in_sqft', 'Age_Category',
           'Floor_Category', 'Building_Height_Category', 'Furnishing', 'Price_in_rupees','Avg_price_rupee_per_sqft', 
           'Bedrooms', 'Bathrooms', 'Covered_parking', 'Open_parking', 'Balcony', 'Sector_Amenity_Score']]

print("We got total : ", data.shape[0], " Rows and ", data.shape[1], " Features.\n")
data.head(3)

## Note: We temporarily keeping Price Density of Properties for future needs! Won't be used in Modeling due to DL concerns.

We got total :  7143  Rows and  16  Features.



Flat Sector_Locality                Locality  Built_up_area_in_sqft  \
0  2 BHK Flat       Sector 86            New Gurugram                    593   
1  2 BHK Flat       Sector 86            New Gurugram                    690   
2  2 BHK Flat      Sector 108  Dwarka Expressway Belt                   1000   

     Age_Category Floor_Category Building_Height_Category      Furnishing  \
0  Newer Property    Upper Floor         Shorter Building  Semi Furnished   
1  Newer Property    Lower Floor          Medium Building     Unfurnished   
2  Newer Property   Middle Floor          Taller Building  Semi Furnished   

   Price_in_rupees  Avg_price_rupee_per_sqft  Bedrooms  Bathrooms  \
0          6800000                   11470.0         2          2   
1          6300000                    9130.0         2          2   
2         15000000                   15000.0         2          2   

   Covered_parking  Open_parking  Balcony  Sector_Amenity_Score  
0                0             1        1                   2.0  
1                0             0        1                   2.0  
2                1             1        3                   0.0

<br><br><br>

---
# **`Removing Outliers from Usefull Data:`**

#### **`I.`** Removing few Sectors or Localitites with less than 5 Property Listings :

In [144]:
sects = data["Sector_Locality"].value_counts()
print(sects.tail(23).to_dict())

{'Sector 78': 4, 'Sector 30': 4, 'Sector 13': 3, 'Ashok Vihar Phase 2': 3, 'Sector 102A': 3, 'Sector 9': 3, 'Sector 37B': 3, 'Sector 23': 2, 'Sector 21': 2, 'Sector 40': 2, 'Patel Nagar': 2, 'Sector 41': 2, 'Sector 46': 1, 'Sadhrana Bamripur': 1, 'Sector 16': 1, 'Palam Farms': 1, 'Sector 27': 1, 'Sector 88': 1, 'Civil Lines': 1, 'Sector 8': 1, 'Sector 72A': 1, 'Sector 17A': 1, 'Sector 114': 1}


In [145]:
data = data[~ data['Sector_Locality'].isin(sects.tail(23).index)].reset_index(drop=True)
data.head()

Flat Sector_Locality                Locality  Built_up_area_in_sqft  \
0  2 BHK Flat       Sector 86            New Gurugram                    593   
1  2 BHK Flat       Sector 86            New Gurugram                    690   
2  2 BHK Flat      Sector 108  Dwarka Expressway Belt                   1000   
3  3 BHK Flat      Sector 102  Dwarka Expressway Belt                    645   
4  2 BHK Flat      Sector 37D             Old Gurgaon                    570   

     Age_Category Floor_Category Building_Height_Category      Furnishing  \
0  Newer Property    Upper Floor         Shorter Building  Semi Furnished   
1  Newer Property    Lower Floor          Medium Building     Unfurnished   
2  Newer Property   Middle Floor          Taller Building  Semi Furnished   
3  Newer Property   Middle Floor         Shorter Building  Semi Furnished   
4  Newer Property    Lower Floor          Medium Building     Unfurnished   

   Price_in_rupees  Avg_price_rupee_per_sqft  Bedrooms  Bathrooms  \
0          6800000                   11470.0         2          2   
1          6300000                    9130.0         2          2   
2         15000000                   15000.0         2          2   
3          7400000                   11470.0         3          2   
4          6600000                   11580.0         2          2   

   Covered_parking  Open_parking  Balcony  Sector_Amenity_Score  
0                0             1        1                  2.00  
1                0             0        1                  2.00  
2                1             1        3                  0.00  
3                1             1        1                  2.00  
4                1             0        3                  2.67

<br><br>

---
#### **`II.`** Removing few Flat Types with less than 5 Property Listings :

In [151]:
flats = data["Flat"].value_counts()
print(flats.tail(3).to_dict())

{'1.5 BHK Flat': 3, '6 BHK Flat': 1, '7 BHK Flat': 1}


In [152]:
data = data[~ data['Flat'].isin(flats.tail(3).index)].reset_index(drop=True)
data.head()

Flat Sector_Locality                Locality  Built_up_area_in_sqft  \
0  2 BHK Flat       Sector 86            New Gurugram                    593   
1  2 BHK Flat       Sector 86            New Gurugram                    690   
2  2 BHK Flat      Sector 108  Dwarka Expressway Belt                   1000   
3  3 BHK Flat      Sector 102  Dwarka Expressway Belt                    645   
4  2 BHK Flat      Sector 37D             Old Gurgaon                    570   

     Age_Category Floor_Category Building_Height_Category      Furnishing  \
0  Newer Property    Upper Floor         Shorter Building  Semi Furnished   
1  Newer Property    Lower Floor          Medium Building     Unfurnished   
2  Newer Property   Middle Floor          Taller Building  Semi Furnished   
3  Newer Property   Middle Floor         Shorter Building  Semi Furnished   
4  Newer Property    Lower Floor          Medium Building     Unfurnished   

   Price_in_rupees  Avg_price_rupee_per_sqft  Bedrooms  Bathrooms  \
0          6800000                   11470.0         2          2   
1          6300000                    9130.0         2          2   
2         15000000                   15000.0         2          2   
3          7400000                   11470.0         3          2   
4          6600000                   11580.0         2          2   

   Covered_parking  Open_parking  Balcony  Sector_Amenity_Score  
0                0             1        1                  2.00  
1                0             0        1                  2.00  
2                1             1        3                  0.00  
3                1             1        1                  2.00  
4                1             0        3                  2.67

<br><br>

In [154]:
print("Finally our data is ready for Model Training & Selection : ", data.shape)

Finally our data is ready for Model Training & Selection :  (7094, 16)


<br><br><br><br>

---
# **`Saving data for Model Building in Separate CSV File:`**

In [168]:
## Creating a separate CSV File with above cleaned data:
data.to_csv("Gurgaon_Real_Estate_Final.csv")

<br><br><br>

<br><br><br>

---
# **`Saving data to build "Recommendation Engine":`**

In [178]:
## Keeping only the usefull features:
data_rec = df[['Flat', 'Sector_Locality', 'Locality',
           'Seller_Builder', 'Sector_Locality_Price_Density', 'EMI',
           'EMI_in_rupees', 'Built_Up_Area', 'Built_up_area_in_sqft', 'Avg_Price',
           'Avg_price_rupee_per_sqft', 'Age_of_property',
           'Age_of_property_in_years', 'Age_Category',
           'Floor', 'Floor_number', 'Building_height', 'Floor_Category',
           'Building_Height_Category', 'Facing', 'Furnishing', 'Society',
           'Brokerage', 'Brokerage_in_rupees', 'Price', 'Price_in_rupees',
           'Bedrooms', 'Bathrooms', 'Parking', 'Covered_parking', 'Open_parking',
           'Balcony', 'Advertised', 'Amenities', 'Nearby_landmarks',
           'Prop_description', 'Link', 'Sector_Amenity_Score']]

print("We got total : ", data_rec.shape[0], " Rows and ", data_rec.shape[1], " Features.\n")
data_rec.head(3)

## Note: We temporarily keeping Price Density of Properties for future needs! Won't be used in Modeling due to DL concerns.

We got total :  7143  Rows and  38  Features.



Flat Sector_Locality                Locality  \
0  2 BHK Flat       Sector 86            New Gurugram   
1  2 BHK Flat       Sector 86            New Gurugram   
2  2 BHK Flat      Sector 108  Dwarka Expressway Belt   

                      Seller_Builder  Sector_Locality_Price_Density  \
0  Pyramid Infratech Private Limited                        10620.0   
1  Pyramid Infratech Private Limited                        10620.0   
2                Experion Developers                        18375.0   

                      EMI  EMI_in_rupees Built_Up_Area  Built_up_area_in_sqft  \
0  EMI starts at ₹36.01 K          36010     593 sq.ft                    593   
1  EMI starts at ₹33.36 K          33360     690 sq.ft                    690   
2  EMI starts at ₹74.47 K          74470    1000 sq.ft                   1000   

        Avg_Price  Avg_price_rupee_per_sqft Age_of_property  \
0  ₹11.47 K/sq.ft                   11470.0     1 Years Old   
1   ₹9.13 K/sq.ft                    9130.0     1 Years Old   
2     ₹15 K/sq.ft                   15000.0      2 Year Old   

  Age_of_property_in_years    Age_Category     Floor Floor_number  \
0                        1  Newer Property  12 of 14           12   
1                        1  Newer Property   4 of 15            4   
2                        2  Newer Property   9 of 26            9   

  Building_height Floor_Category Building_Height_Category             Facing  \
0              14    Upper Floor         Shorter Building        East facing   
1              15    Lower Floor          Medium Building                  -   
2              26   Middle Floor          Taller Building  North-East facing   

       Furnishing                 Society Brokerage  Brokerage_in_rupees  \
0  Semi Furnished           Pyramid Elite    68,000                68000   
1     Unfurnished           Pyramid Elite    63,000                63000   
2  Semi Furnished  Experion The Heartsong  1.5 Lacs               150000   

    Price  Price_in_rupees  Bedrooms  Bathrooms                       Parking  \
0  68.0 L          6800000         2          2                1 Open Parking   
1  63.0 L          6300000         2          2                    No Parking   
2  1.5 Cr         15000000         2          2  1 Covered and 1 Open Parking   

   Covered_parking  Open_parking  Balcony             Advertised  \
0                0             1        1  More than a month ago   
1                0             0        1            17 days ago   
2                1             1        3  More than a month ago   

                                           Amenities  \
0  ['Lift', 'Power Backup', 'Garden', 'Sports', '...   
1                                                NaN   
2  ['Amphitheater', 'Cricket Pitch', 'Gazebo', 'S...   

                                    Nearby_landmarks  \
0  [['School', "St. Xavier's High School"], ['Hos...   
1  [['School', "St. Xavier's High School"], ['Hos...   
2  [['School', 'The Shikshiyan School'], ['Hospit...   

                                    Prop_description  \
0  Looking for a 2 BHK Flat for sale in Gurgaon? ...   
1  Best 2 BHK Flat for modern-day lifestyle is no...   
2  2 BHK Flat for sale in Sector 108, Gurgaon - c...   

                                                Link  Sector_Amenity_Score  
0  https://housing.com/in/buy/resale/page/1761033...                   2.0  
1  https://housing.com/in/buy/resale/page/1681859...                   2.0  
2  https://housing.com/in/buy/resale/page/1767484...                   0.0

In [180]:
## Creating a separate CSV File with above cleaned data:
data_rec.to_csv("Gurgaon_Real_Estate_data_for_recommendation_engine.csv")

<br><br><br>

# **NOW WE CAN MOVE TO MODEL TRAINING & SELECTION...**
---